### PR TITLE
fix(tmux): render base+VS16 emoji at two columns in control-mode panes

### LINKED ISSUE #2590: bug(tmux): VS16 emoji sometimes render 1 cell too narrow in control-mode panes
I reproduced this myself. AI helped organize the write-up and the root-cause hypothesis, but the behavior, repro, and the "it flips between correct and broken under rerender pressure" detail are from my own testing.

## Summary

In tmux sessions viewed through aoe's control-mode pane renderer, emoji formed by a base codepoint + U+FE0F (VS16) sometimes render as width 1 instead of width 2.

When that happens, the rest of the line shifts left by one cell. The bug is especially noticeable in heavily rerendering TUIs such as Claude Code.

Important nuance: this is not a perfectly stable "always broken" state. Under redraw pressure, the same line can flip between:

- correct alignment
- broken alignment
- correct again

So the symptom is intermittent, but the failure shape is consistent: when it breaks, the cell immediately after the VS16 emoji is one column too far left.

## Repro

In an aoe-attached tmux pane:

```sh
printf '[가][😀][❤️][⚠️][✅][🇰🇷]\n'
```

### Expected

All bracket pairs stay aligned.

### Actual

- OK: `가`, `😀` (U+1F600), `✅` (U+2705), `🇰🇷`
- Broken when the bug manifests: `❤️` (U+2764 U+FE0F), `⚠️` (U+26A0 U+FE0F)

When broken, the bracket after `❤️` or `⚠️` shifts left by one cell.

This does not necessarily stay broken continuously. In the same session, and sometimes across repeated redraws of the same UI region, the VS16 emoji may briefly render correctly and then render too narrow again.

## Control comparison

Attaching the same tmux session directly in the host terminal with `tmux attach` renders correctly.

That makes this look like a divergence in aoe's renderer rather than tmux itself.

## Why this looks renderer-specific

- the same tmux content renders correctly when attached directly
- the producing TUI also appears to lay the row out as if the grapheme width is 2
- the visible desync shows up in aoe's control-mode pane rendering layer

## Root-cause hypothesis

This looks like a grapheme-width bug rather than a plain codepoint-width bug.

My guess is that the renderer is using `eastasianwidth`-style codepoint-oriented width logic, which does not account for grapheme-level presentation changes such as:

- U+FE0F (VS16)
- ZWJ sequences
- regional-indicator clustering

That would explain why base characters such as:

- `❤` (U+2764)
- `⚠` (U+26A0)

get treated as width 1 at the codepoint level, while the actual rendered grapheme with U+FE0F is visually width 2.

Meanwhile, tmux and the producing TUI appear to lay out the row as width 2 for those graphemes, so aoe occasionally paints a row that is one cell narrower than the upstream layout expects.

## Environment

- aoe (`cargo install --git github.com/agent-of-empires/agent-of-empires`)
- tmux 3.6b, linked against libutf8proc 2.x
- Host terminal: iTerm2

## Notes

- toggling iTerm2's Unicode-9 width option does not affect the behavior
- that also points away from the host terminal's width table and toward aoe's own renderer

## Suggested fix

Use grapheme-aware width handling instead of codepoint-only width handling.

Possible directions:

- `string-width` + `emoji-regex`
- xterm.js `@xterm/addon-unicode-graphemes`
- another grapheme-aware Unicode width provider

At minimum, `<emoji base> + U+FE0F` should be treated as width 2.

## Acceptance check

This should stay aligned in an aoe-attached tmux pane, even across repeated redraws:

```sh
printf '[가][😀][❤️][⚠️][✅][🇰🇷]\n'
```

It should not alternate between correct and shifted states for the VS16 emoji.

## Screenshots

I don't have a cleaned-up attachment in this report yet, but if useful I can add:

1. a broken aoe frame where the bracket after `❤️` or `⚠️` is visibly shifted left
2. a correct frame of the same content
   - either a later aoe frame where it lines up again
   - or the same tmux session viewed directly via `tmux attach`

### PR DESCRIPTION
**Pull request by Seluj78**
**Note:** Claude handled the implementation and prose via back-and-forth. Idea, decisions, and everything else are mine. — @Seluj78

## Description

Emoji formed by a base codepoint plus U+FE0F (VS16), such as ❤️ (U+2764 U+FE0F) and ⚠️ (U+26A0 U+FE0F), rendered one cell too narrow in aoe's control-mode pane renderer, shifting the following cell left and flipping between correct and broken under redraw pressure (issue #2590).

Root cause: the pane byte stream feeds a `vt100::Parser` grid, and vt100 measures width per codepoint. The base is width 1 and VS16 is a width-0 combining mark, so the grapheme occupies a single grid cell and the cursor ends one column short of what tmux and the producing app assume (both treat it as width 2). Under relative redraws the app then overwrites the trailing column and the row shifts left; absolute repositioning avoids the overwrite, which is why it flips. vt100's per-codepoint width cannot be hooked, and a dependency fork/vendor was out of scope.

Fix, entirely first-party, no dependency changed:

- `src/tmux/vt.rs`: a byte-stream widener (alongside the existing `lf_to_crlf` preprocessing) injects a cursor-forward after a VS16 that follows a width-1 base, so vt100 reserves the second column and relative writes land correctly. It tracks the cursor column across CR/BS/HT and CUP/CHA/CUF/CUB to skip the last-column case, where vt100 already wraps the next char like a real wide char and a nudge would let it overwrite the emoji. `row_to_ansi` owns the reserved blank so it is not serialised as a stray space.
- `web/src/lib/liveTermLines.ts`: the mobile/web live terminal measured widths per codepoint too, so its wrap and cursor math undercounted VS16 emoji; it now folds the second cell onto the base via one-codepoint lookahead.

The native TUI needed no change: ratatui measures each grapheme with `UnicodeWidthStr::width`, which already handles VS16 as width 2.

Closes #2590

## PR Type

- [ ] New Feature
- [x] Bug Fix
- [ ] Refactor
- [ ] Documentation
- [ ] Infrastructure / CI

## How to test

- [ ] In an aoe-attached tmux pane, run `printf '[가][😀][❤️][⚠️][✅][🇰🇷]\n'` and confirm every bracket pair stays aligned.
- [ ] Open a heavily-redrawing TUI (e.g. Claude Code) in an aoe pane, exercise redraws, and confirm the cell after ❤️ / ⚠️ no longer flips between correct and shifted.
- [ ] View the same session in the web/mobile live terminal and confirm VS16 emoji render two cells wide with the following text aligned.
- [ ] Confirm plain (text-presentation) symbols without VS16, e.g. a bare `❤`, still render one cell wide.

## Checklist

- [x] New and existing tests pass
- [x] Documentation was updated where necessary
- [ ] For UI changes: included screenshot or recording

## Test Coverage Analysis

**Web dashboard / structured view (Playwright user story)**
- [ ] N/A: this PR doesn't change a user-facing dashboard flow
- [ ] Added or updated a Playwright spec under `web/tests/` and updated `web/tests/coverage-matrix.json`
- [x] Skipped a Playwright test, because: the change is pure cell-width logic in `liveTermLines`, covered by the Vitest unit spec `web/src/lib/liveTermLines.test.ts` (already mapped in `web/tests/coverage-matrix.json`), which is the right altitude for width math.

**TUI / CLI (e2e test)**
- [ ] N/A: this PR doesn't change TUI rendering, a CLI subcommand, or session lifecycle
- [ ] Added or updated an e2e test under `tests/e2e/`
- [x] Skipped an e2e test, because: the grid/serialisation path is covered by deterministic unit tests in `src/tmux/vt.rs` (`vs16_*`) that feed bytes through `vt100::Parser` + `row_to_ansi` and assert exact columns; a tmux-driven e2e would depend on the terminal font's emoji width and be nondeterministic.

## AI Usage

- [ ] No AI was used
- [x] AI was used

**AI Model/Tool used:** Claude (via Claude Code, agent-of-empires `/aoe-implement` skill)

**Any Additional AI Details you'd like to share:**
Investigation, the width-provider debate, and implementation were drafted by Claude under my direction. I made the design calls (first-party byte-stream fix over a vt100 fork or vendor) and reviewed each commit.

- [x] I am an AI Agent filling out this form (check box if true)

---
<a href="https://app.blacksmith.sh/agent-of-empires/codesmith/agent-of-empires/pr/2597"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1785489023&installation_id=139138837&pr_number=2597&repository=agent-of-empires%2Fagent-of-empires&return_to=https%3A%2F%2Fgithub.com%2Fagent-of-empires%2Fagent-of-empires%2Fpull%2F2597&signature=1412f0aff3d0986f58381347754bb2e6fe1d0f5e49cc77b93283dcd06e1cf08b"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

## What changed
- Fixed tmux control-mode pane rendering so emoji made from a width-1 base character followed by VS16 (e.g., ❤️, ⚠️) reserve **two terminal cells** instead of one.
- Prevented the reserved second cell from being emitted as an actual visible character/blank in the serialized output.
- Updated the web client’s live terminal width + wrapping logic to measure VS16 sequences consistently (one-codepoint lookahead).
- Added/expanded unit tests to cover VS16 width, wrapping behavior, and redraw/margin edge cases (including resync after seeding/resizing).

## Benefit
- VS16 emoji stay aligned and do not cause the following cell to shift left.
- Rendering stays stable under repeated redraws, keeping aoe panes consistent with direct `tmux attach`.

## Technical notes
- `src/tmux/vt.rs`: introduced a byte-stream “VS16 widener” that tracks cursor/columns across chunks and injects a cursor-forward nudge when VS16 needs a reserved second cell; `row_to_ansi` ensures the reserved blank is only owned/emitted when truly empty/unmodified.
- `web/src/lib/liveTermLines.ts`: adjusted width/wrapping to look ahead by code point so VS16 presentation sequences are treated as width-2 units.

### CODE REVIEW

File: src/tmux/vt.rs [hunk 1]
```diff
@@ -419,12 +615,20 @@ fn row_to_ansi(screen: &vt100::Screen, row: u16, cols: u16) -> String {
...
+            // columns so that reserved blank is not serialised as a stray space
+            // and the following cell keeps its column.
+            if advance == 1 && cell.contents().ends_with('\u{fe0f}') {
+                advance = 2;
+            }
         } else {
             out.push(' ');
         }
-        col += if cell.is_wide() { 2 } else { 1 };
+        col += advance;
```
[the remark this answers is not in the document]
Comment (Seluj78) [reply src/tmux/vt.rs:650]: Good catch. Fixed in https://github.com/agent-of-empires/agent-of-empires/pull/2597/commits/9395a745: the advance to two columns is now gated on that column actually being the reserved blank (empty, unstyled, and `col + 1 < last`), so a later absolute write or styled fill there is emitted instead of swallowed. Added a regression test (`vs16_does_not_swallow_real_content_in_the_next_cell`).

File: web/src/lib/liveTermLines.ts [hunk 2]
```diff
@@ -46,10 +46,19 @@ function cellWidth(codePoint: string): number {
...
+}
+
 function textWidth(text: string): number {
   if (ASCII_PRINTABLE_ONLY.test(text)) return text.length;
+  const cps = [...text];
   let width = 0;
-  for (const ch of text) width += cellWidth(ch);
+  for (let i = 0; i < cps.length; i++) width += cellWidthWithNext(cps[i]!, cps[i + 1]);
   return width;
 }
```
[the remark this answers is not in the document]
Comment (Seluj78) [reply web/src/lib/liveTermLines.ts:63]: This isn't reachable through aoe's own pipeline: a base+VS16 grapheme occupies a single vt100 grid cell, and `row_to_ansi` emits both codepoints contiguously under one SGR run, so `parseAnsi` never splits them across segments (no SGR is injected mid-grapheme). Rather than carry cross-segment lookbehind state for an input shape the serializer can't produce, I'll leave the single-string lookahead as is. Happy to revisit if we ever feed externally-authored ANSI through here.

**Comment by Seluj78** [comment]
> Synchronize the widener with seeded and resized parser state. (@coderabbitai, outside-diff)

Fixed in https://github.com/agent-of-empires/agent-of-empires/pull/2597/commits/dcdd539b: the live `Vs16Widener` now resyncs `col` and `cols` from the parser's cursor and screen size under the lock before each chunk, so the last-column gate reflects the post-seed cursor and any resize instead of stale arm-time state. Added `vs16_resync_drives_the_margin_gate`.

**Comment by Seluj78** [comment]
Closing: root cause is a tmux 3.6b wide-character redraw bug (fixed in tmux 3.7, "Correct redrawing of wide characters when overwritten"), not aoe. Verified aoe renders base+VS16 correctly on tmux 3.7. See #2590 for the investigation.
